<?xml version="1.0"?>
<robot xmlns:xacro="http://ros.org/wiki/xacro">
  <xacro:macro name="psen_scan" params="name ros_topic">
    <xacro:property name="min_range" value="0.1" />
    <xacro:property name="max_range" value="4.0" />
    <xacro:property name="min_angle" value="3.0" />
    <xacro:property name="max_angle" value="0.1" />

    <xacro:property name="mass" value="0.5" /> 
    <xacro:property name="radius" value="0.1" />
    <xacro:property name="height" value="0.2" /> 
    <xacro:property name="z_offset" value="0.2" /> 
    <!-- TODO correct dimensions -->
    
    <!-- Location of mounting point at the lower back of device -->
    <link name="${name}_mount_link">
      <inertial>
        <mass value="0.001" />
        <origin xyz="0 0 0" />
        <inertia ixx="0.000001" ixy="0.0" ixz="0.0"
          iyy="0.000001" iyz="0.0" izz="0.000001" />
      </inertial>
    </link>

    <joint name="${name}_mount_joint" type="fixed">
      <parent link="${name}_mount_link" />
      <child link="${name}" />
      <origin rpy="0 0 0" xyz="0 0 ${z_offset}"/>
      <!-- TODO correct dimensions -->
    </joint>

    <!-- Location of scanner body -->
    <link name="${name}">
      <visual>
        <geometry>
            <geometry type="cylinder" radius="${radius}" length="${height}" />
        </geometry>
        <material name="yellow" >
          <color rgba="0 0 1 1" />
        </material>
      </visual>
      <collision>
        <geometry>
            <geometry type="cylinder" radius="${radius}" length="${height}" />
        </geometry>
      </collision>
      <inertial>
        <mass value="${mass}" />
        <origin xyz="0 0 -0.026" />
        <inertia ixx="${0.0833333 * mass * (radius * radius + height * height)}" ixy="0.0" ixz="0.0"
          iyy="${0.0833333 * mass * (radius * radius + height * height)}" iyz="0.0"
          izz="${0.0833333 * mass * (radius * radius + radius * radius)}" />
      </inertial>
    </link>
    <xacro:psen_scan_gazebo_v0 name="${name}" link="${name}" ros_topic="${ros_topic}" update_rate="15.0" min_angle="${min_angle}" max_angle="${max_angle}" min_range="${min_range}" max_range="${max_range}"/>

    <joint name="${name}_scan_joint" type="fixed">
      <parent link="${name}" />
      <child link="${name}_scan" />
      <origin rpy="0 ${pi} 0" xyz="0 0 0"/>
    </joint>
    
    <!-- This is the location of the laser scan -->
    <link name="${name}_scan">
    </link>
  </xacro:macro>


  <xacro:macro name="psen_scan_gazebo_v0" params="name link ros_topic update_rate min_angle max_angle min_range max_range">
    <gazebo reference="${link}">
      <material value="Gazebo/Yellow" />
      <sensor type="ray" name="${name}">
        <always_on>true</always_on>
        <update_rate>${update_rate}</update_rate>
        <pose>0 0 0 0 0 0</pose>
        <visualize>false</visualize>
        <ray>
          <scan>
            <horizontal>
              <samples>271</samples>
              <resolution>1</resolution>
              <min_angle>${min_angle}</min_angle>
              <max_angle>${max_angle}</max_angle>
            </horizontal>
          </scan>
          <range>
            <min>${min_range}</min>
            <max>${max_range}</max>
            <resolution>0.01</resolution>
          </range>
        </ray>
        <plugin name="gazebo_ros_${name}_controller" filename="libgazebo_ros_laser.so">
          <gaussianNoise>0.005</gaussianNoise>
          <alwaysOn>true</alwaysOn>
          <updateRate>${update_rate}</updateRate>
          <topicName>${ros_topic}</topicName>
          <frameName>${link}</frameName>
        </plugin>
      </sensor>
    </gazebo>
  </xacro:macro>
</robot>